Ensure that createViewSession preserves the trace token

huw0 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information 
at https://trino.io/development/process.html, 
at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md 
and contact us on #core-dev in Slack. -->
<!-- Provide an overview for maintainers and reviewers. -->
## Description

`Session`.`createViewSession` creates a new Session that is used when analysing views.

Presently this does not pass through the trace token. This is breaking the use of views with our custom connectors which mandate that this token is populated.

Therefore this is a very simple PR to ensure that the trace token is populated in `createViewSession` if it is present on the outer session.

<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
## General
* createViewSession should preserve the client trace token ({issue}`issuenumber`)
```

Praveen2112 commented on core/trino-main/src/main/java/io/trino/Session.java:
```diff
@@ -661,6 +661,7 @@ public Session createViewSession(Optional<String> catalog, Optional<String> sche
                 .setRemoteUserAddress(getRemoteUserAddress().orElse(null))
                 .setUserAgent(getUserAgent().orElse(null))
                 .setClientInfo(getClientInfo().orElse(null))
+                .setTraceToken(traceToken)
```
nit:  `.setTraceToken(getTraceToken())`

huw0 replied:
Thanks, I should've caught that.

dain commented:
This looks good to me, but if a test is possible, please add that.  If not (or really annoying), please let us know, and one of us will merge it as is.

Praveen2112 commented:
@huw0 For testing I think we could use `MockConnectorMetadata` which could make use of the trace token at the time of its execution.

huw0 commented:
I've added some unit testing around `createViewSession` and rebased the branch.
There seem to be some test failures caused by github rate limiting and some random plugin failures that change every time. Expect these should pass once any wider issues resolve.

Praveen2112 commented:
Instead of this test - WDYT about https://github.com/trinodb/trino/pull/25716#issuecomment-2853389163 ?

huw0 commented:
@Praveen2112 - I've added an additional commit with an alternative test using the approach you've outlined.

huw0 commented:
> I'd be in favour of dropping this 3rd commit and sticking with the first two. The unit test alone increases the existing coverage.

@dain @Praveen2112 - Please see comment above. I'd like to get this merged soon if there are no objections?

raunaqmorarka commented on core/trino-main/src/test/java/io/trino/TestSession.java:
```diff
@@ -61,4 +69,53 @@ public void testAddSecondCatalogProperty()
+                .build();
+
+        Session viewSession = originalSession.createViewSession(catalog, schema, identity,
+                Collections.emptyList());
```
`Collections.emptyList()` -> `ImmutableList.of()`
Please keep all arguments on same line, or each argument on new line

raunaqmorarka commented on core/trino-main/src/test/java/io/trino/TestSession.java:
```diff
@@ -61,4 +69,53 @@ public void testAddSecondCatalogProperty()
+    @Test
+    public void testCreateViewSession()
+    {
+        final var catalog = Optional.of("test_catalog");
```
We don't declare local variables as `final`

raunaqmorarka commented on core/trino-main/src/test/java/io/trino/TestSession.java:
```diff
@@ -61,4 +69,53 @@ public void testAddSecondCatalogProperty()
+    public void testCreateViewSession()
+    {
+        final var catalog = Optional.of("test_catalog");
+        final var schema = Optional.of("test_schema");
```
We avoid using `var`.
Please look at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#code-style to get an idea of the overall development guidelines
